swarm/storage/mru/updateheader.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package mru
+
+import (
+	"github.com/ethereum/go-ethereum/swarm/log"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+// updateHeader models the non-payload components of a Resource Update
+type updateHeader struct {
+	UpdateLookup        // UpdateLookup contains the information required to locate this resource (components of the search key used to find it)
+	multihash    bool   // Whether the data in this Resource Update should be interpreted as multihash
+	metaHash     []byte // SHA3 hash of the metadata chunk (less ownerAddr). Used to prove ownerhsip of the resource.
+}
+
+const metaHashLength = 32
+
+// updateLookupLength bytes
+// 1 byte flags (multihash bool for now)
+// 32 bytes metaHash
+const updateHeaderLength = updateLookupLength + 1 + metaHashLength
+
+// binaryPut serializes the resource header information into the given slice
+func (h *updateHeader) binaryPut(serializedData []byte) error {
+	if len(serializedData) != updateHeaderLength {
+		return NewErrorf(ErrInvalidValue, "Incorrect slice size to serialize updateHeaderLength. Expected %d, got %d", updateHeaderLength, len(serializedData))
+	}
+	if len(h.metaHash) != metaHashLength {
+		log.Warn("Call to updateHeader.binaryPut with incorrect metaHash")
+		return NewError(ErrInvalidValue, "updateHeader.binaryPut called without metaHash set")
+	}
+	if err := h.UpdateLookup.binaryPut(serializedData[:updateLookupLength]); err != nil {
+		return err
+	}
+	cursor := updateLookupLength
+	copy(serializedData[cursor:], h.metaHash[:metaHashLength])
+	cursor += metaHashLength
+
+	if h.multihash {
+		serializedData[cursor] = 0x01
+	}
+	cursor++
```

Comment:
remove please

Reply:
I don't agree. A lower level protocol must not have to inspect the higher level payload to figure out its own frame length. Previous code was deducting frame length out of the multihash length.

As part of the serialization refactor effort, it was a [CF](https://www.youtube.com/embed/SlA9hmrC8DU?start=96&end=98&autoplay=1)<sup>(TM)</sup> to keep using "data length" = 0 to indicate a multihash and then go to great lengths of [interesting code](https://github.com/ethersphere/go-ethereum/blob/d0e18ba9db9a21cc1ad5ca8db24c427f78d57626/swarm/storage/mru/resource.go#L591) to figure out the actual length of the data. :-)

I advocate for a clean "flags" field with space for 8 flags, flag 0 being "ismultihash".

Reply:
I meant the cursor increment...

Reply:
> I meant the cursor increment...

Ah... eh... 😅 oops!, heh, sorry about that Louis.

that `cursor++` is left there on purpose. You'll see this pattern in the other serializers. The compiler will throw it away because it leads nowhere, but helps make sure that a future developer  that adds something does not forget the cursor should have incremented after that field was added. Think of it as documentation, the cursor increment goes together with the block that interacts with the slice.

Also, if someone moves code around (imagine we wanted the flags field to be the first one), they won't mess serialization up.

Reply:
> The compiler will throw it away because it leads nowhere

I didn't know that. Fine, then :)